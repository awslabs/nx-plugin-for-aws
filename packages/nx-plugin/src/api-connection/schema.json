{
  "$schema": "https://json-schema.org/schema",
  "$id": "APIConnection",
  "title": "",
  "type": "object",
  "properties": {
    "sourceProject": {
      "type": "string",
      "description": "The source project which will call the API",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the source project which will call the API",
      "x-dropdown": "projects"
    },
    "targetProject": {
      "type": "string",
      "description": "The target project containing your API",
      "x-prompt": "Select the target project containing your API",
      "x-dropdown": "projects"
    }
  },
  "required": ["sourceProject", "targetProject"]
}
