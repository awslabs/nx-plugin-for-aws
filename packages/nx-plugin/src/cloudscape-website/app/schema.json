{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"],
  "examplesFile": "../../../docs/application-examples.md"
}
