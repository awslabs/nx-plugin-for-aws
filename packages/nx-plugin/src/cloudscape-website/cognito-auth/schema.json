{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    },
    {
      "command": "nx g app myapp --classComponent",
      "description": "Use class components instead of functional components"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "the root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "cognitoDomain": {
      "description": "The cognito domain prefix to use",
      "type": "string",
      "x-priority": "important",
      "x-prompt": "the cognito domain prefix to use i.e: https://{{cognitoDomain}}-123456789.auth.ap-southeast-2.amazoncognito.com"
    },
    "allowSignup": {
      "description": "Whether to allow self-signup",
      "type": "boolean",
      "default": false,
      "x-prompt": "Enable self signup?",
      "x-priority": "important"
    }
  },
  "required": ["project", "cognitoDomain"]
}
