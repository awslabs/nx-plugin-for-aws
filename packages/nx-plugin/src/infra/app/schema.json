{
  "$schema": "https://json-schema.org/schema",
  "$id": "Infra",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "ruleSet": {
      "type": "string",
      "enum": [
        "aws_prototyping",
        "cfn_nag",
        "hipaa",
        "nist_csf",
        "pci_dss_3_2_1",
        "well_architected_reliability",
        "well_architected_security"
      ],
      "x-priority": "important",
      "x-prompt": "What rule set do you want to validate your AWS resources with?",
      "description": "Rule set to validate your AWS resources with.",
      "default": "aws_prototyping"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"]
}
