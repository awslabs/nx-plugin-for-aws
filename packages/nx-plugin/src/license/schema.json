{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "License",
  "title": "Add LICENSE files, and configure source code license headers",
  "type": "object",
  "properties": {
    "license": {
      "type": "string",
      "description": "SPDX License Identifier",
      "enum": ["Apache-2.0", "MIT", "ASL"],
      "default": "Apache-2.0",
      "x-priority": "important"
    },
    "copyrightHolder": {
      "type": "string",
      "description": "Copyright holder name",
      "default": "Amazon.com, Inc. or its affiliates",
      "x-priority": "important"
    }
  },
  "required": []
}
