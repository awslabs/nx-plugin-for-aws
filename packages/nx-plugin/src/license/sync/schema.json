{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "license-headers",
  "title": "Sync generator for adding license headers to source code",
  "type": "object",
  "properties": {},
  "required": []
}
