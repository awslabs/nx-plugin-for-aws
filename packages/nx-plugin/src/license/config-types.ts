/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
export interface LicenseConfig {
  /**
   * Configuration for source code license headers
   * Omit if you do not want any source code license headers to be synced
   */
  header?: LicenseHeaderConfig;
  /**
   * Configuration for LICENSE files
   */
  files?: LicenseFileConfig;
}

/**
 * Configuration for the LICENSE file sync
 */
export interface LicenseFileConfig {
  /**
   * Exclude LICENSE file sync checks for the given glob patterns
   */
  exclude?: string[];
}

/**
 * Specifies license content as a path to a file
 */
export interface LicenseFilePathContent {
  /**
   * Load the license text from the specified file (relative to the workspace root)
   */
  filePath: string;
}

/**
 * Specifies license content inline
 */
export interface LicenseLinesContent {
  /**
   * Inline license text
   */
  lines: string[];
}

/**
 * Configuration for source code license header sync
 */
export interface LicenseHeaderConfig {
  /**
   * Content of the license header
   */
  content: LicenseFilePathContent | LicenseLinesContent;
  /**
   * Specifies the license header format for given glob patterns
   */
  format: {
    [filePattern: string]: LicenseHeaderFormat;
  };
  /**
   * Exclude out of sync source code license header checks for the given glob patterns
   */
  exclude?: string[];
  /**
   * User-specified comment syntax for file types we don't have config for
   */
  commentSyntax?: {
    [ext: string]: CommentSyntax;
  };
}

/**
 * Format settings for source code license headers
 */
export interface LicenseHeaderFormat {
  /**
   * Text written to the line prior to the license lines
   * Use this to start a block comment if desired
   */
  blockStart?: string;
  /**
   * Text written to the start of each line of the license
   */
  lineStart?: string;
  /**
   * Text written to the end of each line of the license
   */
  lineEnd?: string;
  /**
   * Text written to the line after the license lines
   * Use this to end a block comment if desired
   */
  blockEnd?: string;
}

/**
 * Defines comment syntax for a particular language
 * Used to instruct the license sync generator to find the first comment in a file
 */
export interface CommentSyntax {
  /**
   * If the language supports line comments, specify the character(s) used to start a line comment
   */
  line?: string;
  /**
   * If the language supports block comments, specify the character(s) used to start and end the block comment
   */
  block?: {
    start: string;
    end: string;
  };
}
