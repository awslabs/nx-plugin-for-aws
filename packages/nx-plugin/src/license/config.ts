/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { SPDXLicenseIdentifier } from './schema';
import {
  readAwsNxPluginConfig,
  updateAwsNxPluginConfig,
} from '../utils/config/utils';
import { CommentSyntax, LicenseConfig } from './config-types';

/**
 * Licenses containing a placeholder for the copyright year
 */
const LICENSES_WITH_YEAR: Set<SPDXLicenseIdentifier> = new Set(['MIT']);

/**
 * Defines the comment syntax for popular programming languages
 */
export const LANGUAGE_COMMENT_SYNTAX: { [ext: string]: CommentSyntax } = {
  // C++ style comments
  ...Object.fromEntries(
    [
      // Node
      'js',
      'ts',
      'cjs',
      'mjs',
      'cts',
      'mts',
      // React
      'jsx',
      'tsx',
      // Web
      'sass',
      'scss',
      'php',
      // Java
      'java',
      // Smithy
      'smithy',
      // TypeSpec
      'tsp',
      // Golang
      'go',
      // Rust
      'rs',
      // C and C++
      'c',
      'cpp',
      'cxx',
      'cc',
      'h',
      'hpp',
      // C#
      'cs',
      // Scala
      'scala',
      'sc',
      'sbt',
      // Kotlin
      'kt',
      // Swift
      'swift',
    ].map((ext) => [ext, { line: '//', block: { start: '/*', end: '*/' } }]),
  ),
  // Bash style comments
  ...Object.fromEntries(
    [
      // Bash
      'sh',
      // Perl
      'pl',
      // YAML
      'yaml',
      'yml',
      // R
      'r',
    ].map((ext) => [ext, { line: '#' }]),
  ),
  // Docker
  Dockerfile: { line: '#' },
  // Python
  py: { line: '#', block: { start: '"""', end: '"""' } },
  // Web
  html: { block: { start: '<!--', end: '-->' } },
  css: { block: { start: '/*', end: '*/' } },
  // Ruby
  rb: { line: '#', block: { start: '=begin', end: '=end' } },
  // F#
  ...Object.fromEntries(
    ['fs', 'fsx'].map((ext) => [
      ext,
      { line: '//', block: { start: '(*', end: '*)' } },
    ]),
  ),
  // Visual Basic
  vb: { line: "' " },
  // Lua
  lua: { line: '--', block: { start: '--[[', end: ']]' } },
  // PowerShell
  ps1: { line: '#', block: { start: '<#', end: '#>' } },
  psm1: { line: '#', block: { start: '<#', end: '#>' } },
  // Markdown
  md: { block: { start: '<!--', end: '-->' } },
};

/**
 * Build the default license config for a given license
 */
export const defaultLicenseConfig = (
  spdx: SPDXLicenseIdentifier,
  copyrightHolder: string,
): LicenseConfig => {
  switch (spdx) {
    case 'ASL': {
      // For ASL, we use a "box" style license header
      const rawContent = [
        `Copyright ${copyrightHolder}. All Rights Reserved.`,
        '',
        'Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance',
        'with the License. A copy of the License is located at',
        '',
        '   https://aws.amazon.com/asl/',
        '',
        'or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES',
        'OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions',
        'and limitations under the License.',
      ];
      const maxLen = Math.max(...rawContent.map((line) => line.length));
      const lines = rawContent.map(
        (line) => `${line}${' '.repeat(maxLen - line.length)}`,
      );
      return {
        spdx,
        copyrightHolder,
        header: {
          content: { lines },
          format: {
            '**/*.{js,ts,jsx,tsx,mjs,mts}': {
              blockStart: `/***${'*'.repeat(maxLen)}**`,
              lineStart: ' *  ',
              lineEnd: ' *',
              blockEnd: ` ***${'*'.repeat(maxLen)}**/`,
            },
            '**/*.{py,sh}': {
              blockStart: `###${'#'.repeat(maxLen)}##`,
              lineStart: '#  ',
              lineEnd: ' #',
              blockEnd: `###${'#'.repeat(maxLen)}##`,
            },
          },
          exclude: ['**/*.gen.*'],
        },
      };
    }
    default: {
      return {
        spdx,
        copyrightHolder,
        ...(LICENSES_WITH_YEAR.has(spdx)
          ? {
              copyrightYear: new Date().getFullYear(),
            }
          : {}),
        header: {
          content: {
            lines: [
              `Copyright ${copyrightHolder}. All Rights Reserved.`,
              `SPDX-License-Identifier: ${spdx}`,
            ],
          },
          format: {
            '**/*.{js,ts,jsx,tsx,mjs,mts}': {
              blockStart: '/**',
              lineStart: ' * ',
              blockEnd: ' */',
            },
            '**/*.{py,sh}': {
              blockStart: '#',
              lineStart: '# ',
              blockEnd: '#',
            },
          },
          exclude: ['**/*.gen.*'],
        },
      };
    }
  }
};

/**
 * Write license configuration to the aws nx plugin config
 */
export const writeLicenseConfig = async (tree: Tree, config: LicenseConfig) => {
  await updateAwsNxPluginConfig(tree, {
    license: config,
  });
};

/**
 * Read license configuration
 */
export const readLicenseConfig = async (
  tree: Tree,
): Promise<LicenseConfig | undefined> => {
  return (await readAwsNxPluginConfig(tree))?.license;
};
