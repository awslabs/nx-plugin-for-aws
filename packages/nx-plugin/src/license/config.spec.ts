/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import {
  defaultLicenseConfig,
  readLicenseConfig,
  writeLicenseConfig,
} from './config';
import { SPDXLicenseIdentifier } from './schema';
import { createTreeUsingTsSolutionSetup } from '../utils/test';
import { AWS_NX_PLUGIN_CONFIG_FILE_NAME } from '../utils/config/utils';
import { LicenseConfig } from './config-types';

const LICENSES: SPDXLicenseIdentifier[] = ['Apache-2.0', 'MIT', 'ASL'];

describe('license config', () => {
  let tree: Tree;

  const sampleConfig: LicenseConfig = {
    spdx: 'ASL',
    copyrightHolder: 'Test Inc.',
    header: {
      content: {
        lines: ['this is a test license header'],
      },
      format: {
        '**/*.js': {
          lineStart: '// ',
        },
      },
    },
  };

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  describe('defaultLicenseConfig', () => {
    it.each(LICENSES)(
      'should generate default license config for %s',
      (spdx) => {
        expect(
          defaultLicenseConfig(spdx, 'Test Inc. or its affiliates'),
        ).toMatchSnapshot();
      },
    );
  });

  describe('readLicenseConfig', () => {
    it('should read license configuration', async () => {
      tree.write(
        AWS_NX_PLUGIN_CONFIG_FILE_NAME,
        `
        export default {
          license: ${JSON.stringify(sampleConfig)}
        };
      `,
      );

      expect(await readLicenseConfig(tree)).toEqual(sampleConfig);
    });
  });

  describe('writeLicenseConfig', () => {
    it('should write license configuration', async () => {
      tree.write(AWS_NX_PLUGIN_CONFIG_FILE_NAME, `export default {}`);

      await writeLicenseConfig(tree, sampleConfig);

      expect(tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8')).toContain(
        'this is a test license header',
      );
    });
  });
});
