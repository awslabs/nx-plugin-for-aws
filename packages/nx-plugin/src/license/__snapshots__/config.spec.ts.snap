// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`license config > defaultLicenseConfig > should generate default license config for ASL 1`] = `
{
  "copyrightHolder": "Test Inc. or its affiliates",
  "header": {
    "content": {
      "lines": [
        "Copyright Test Inc. or its affiliates. All Rights Reserved.                                                      ",
        "                                                                                                                 ",
        "Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance       ",
        "with the License. A copy of the License is located at                                                            ",
        "                                                                                                                 ",
        "   https://aws.amazon.com/asl/                                                                                   ",
        "                                                                                                                 ",
        "or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES",
        "OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions   ",
        "and limitations under the License.                                                                               ",
      ],
    },
    "exclude": [],
    "format": {
      "**/*.{js,ts,jsx,tsx,mjs,mts}": {
        "blockEnd": " **********************************************************************************************************************/",
        "blockStart": "/**********************************************************************************************************************",
        "lineEnd": " *",
        "lineStart": " *  ",
      },
      "**/*.{py,sh}": {
        "blockEnd": "######################################################################################################################",
        "blockStart": "######################################################################################################################",
        "lineEnd": " #",
        "lineStart": "#  ",
      },
    },
  },
  "spdx": "ASL",
}
`;

exports[`license config > defaultLicenseConfig > should generate default license config for Apache-2.0 1`] = `
{
  "copyrightHolder": "Test Inc. or its affiliates",
  "header": {
    "content": {
      "lines": [
        "Copyright Test Inc. or its affiliates. All Rights Reserved.",
        "SPDX-License-Identifier: Apache-2.0",
      ],
    },
    "exclude": [],
    "format": {
      "**/*.{js,ts,jsx,tsx,mjs,mts}": {
        "blockEnd": " */",
        "blockStart": "/**",
        "lineStart": " * ",
      },
      "**/*.{py,sh}": {
        "blockEnd": "# ",
        "blockStart": "# ",
        "lineStart": "# ",
      },
    },
  },
  "spdx": "Apache-2.0",
}
`;

exports[`license config > defaultLicenseConfig > should generate default license config for MIT 1`] = `
{
  "copyrightHolder": "Test Inc. or its affiliates",
  "copyrightYear": 2025,
  "header": {
    "content": {
      "lines": [
        "Copyright Test Inc. or its affiliates. All Rights Reserved.",
        "SPDX-License-Identifier: MIT",
      ],
    },
    "exclude": [],
    "format": {
      "**/*.{js,ts,jsx,tsx,mjs,mts}": {
        "blockEnd": " */",
        "blockStart": "/**",
        "lineStart": " * ",
      },
      "**/*.{py,sh}": {
        "blockEnd": "# ",
        "blockStart": "# ",
        "lineStart": "# ",
      },
    },
  },
  "spdx": "MIT",
}
`;
